classes/task/merge_user_accounts.php

```diff
@@ -0,0 +1,198 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+namespace tool_mergeusers\task;
+defined('MOODLE_INTERNAL') || die();
+require_once(__DIR__ . '/../../lib/autoload.php');
+use MergeUserTool;
+use \tool_mergeusers\merge_request;
+/**
+ * Adhoc task to merge user accounts.
+ *
+ * @package   tool_mergeusers
+ * @author    Jordi Pujol-Ahulló <jordi.pujol@urv.cat>
+ * @author    Nicola Vallinoto <n.vallinoto@liguriadigitale.it>
+ * @copyright 2013 Servei de Recursos Educatius (http://www.sre.urv.cat)
+ * @copyright 2023 Liguria Digitale (http://www.liguriadigitale.it)
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * */
+class merge_user_accounts extends \core\task\adhoc_task {
+    /**
+     * Execute the task of merge users accounts.
+     */
+    public function execute() {
+        global $DB;
+        $data = $this->get_custom_data();
+        $maxattempts = get_config('tool_mergeusers', 'maxattempts');
+        $mergerequest = $DB->get_record(merge_request::TABLE_MERGE_REQUEST,
+                                            ['id' => $data->mergerequestid]);
+        if (!$mergerequest) {
+            // Merge request was removed before running this task.
+            return;
+        }
```

Comment:
Imagine a merge request with certain data that never fits, like a username that does not exist. It will keep failing forever on `verify_users_to_keep_and_remove()`, without considering the number of trials.

So, we after having an existing `$mergerequest` record, should evaluate the number of trials, so that:

```php
if ($mergerequest->retries > $maxattempts) {
    // TODO: add a log line that in the $mergerequest->logs array to explain why it is finished: "Reached the number of maximum attempts '3'" or similar.
    $this->update_status_and_log_in_table($mergerequest->id,
                                            merge_request::COMPLETED_WITH_ERRORS,
                                            $mergerequest->logs);
}
```

or something similar.

Reply:
maybe could be useful to add a check inside find_user_id_or_fail. let's think how to do.

Reply:
you already implemented this just after having the mergerequest instance.